Memory snapshot support for pointers and structs [blocks: #4482]

xbauch opened the pull request:
This PR extends the memory snapshot harness to correctly work with pointers and structs (and unions). This requires two functionalities to cooperate: 1) the `memory-analyzer` should produce a JSON file containing all information needed for; 2) the `goto-harness` should be able to read the pointer/struct symbols and correctly instantiate them. Regression tests automate the generation of JSON symbol tables for the harness so that they do not have to be included. The first commit is a squash of the prerequisite PR #4261 (no need to review it here).

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

xbauch commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -210,10 +227,16 @@ code_blockt memory_snapshot_harness_generatort::add_assignments_to_globals(
+    const symbolt &symbol =
+      (symbol_table.lookup(snapshot_symbol.base_name) != nullptr)
+        ? snapshot_symbol
+        : fresh_symbol_copy(snapshot_symbol, symbol_table);
```
The `memory-analyzer` produces a symtab with `tmp` variables that the malloc(ed) pointers point to. These new variables do not exists in the "local" symbol table. So I add them here.

sonodtt replied:
The comment in the code is not so clear on this point, although it mentions it.

I would prefer some text similar to the above specifically mentioning this within the context of memory-analyzer.

hannes-steffenhagen-diffblue reviewed (commented):
Changes look sensible enough to me

tautschnig commented on src/util/std_types.h:
```diff
@@ -25,6 +25,13 @@ Author: Daniel Kroening, kroening@kroening.com
+inline bool is_constant(const typet &type)
+{
+  return type.id() == ID_constant;
+}
```
This makes no sense. If this ever were to be true, we'd have messed something up. Maybe `return type.get_bool(ID_C_constant);`?

tautschnig commented on src/util/std_types.h:
```diff
@@ -1082,6 +1089,13 @@ inline bool can_cast_type<bitvector_typet>(const typet &type)
+inline bool is_signed_or_unsigned_bitvector(const typet &type)
+{
+  return type.id() == ID_signedbv || type.id() == ID_unsignedbv;
+}
```
By now I'd suggest to make this consistent with `integer_bitvector_type`, doing something like `is_integer_bitvector_type`.

tautschnig commented on src/memory-analyzer/Makefile:
```diff
@@ -0,0 +1,37 @@
+SRC = analyze_symbol.cpp\
+			gdb_api.cpp \
+			memory_analyzer_main.cpp \
+			memory_analyzer_parse_options.cpp
```
Looks like a mix of tabs and spaces?

tautschnig commented on src/goto-harness/memory_snapshot_harness_generator.h:
```diff
@@ -117,6 +117,10 @@ class memory_snapshot_harness_generatort : public goto_harness_generatort
   /// \param goto_model: Model where the modification takes place
   void add_init_section(goto_modelt &goto_model) const;
 
+  const symbolt &fresh_symbol_copy(
```
Please add documentation.

xbauch replied:
@tautschnig ok, but the above comments should probably be reflected on the prerequisite PR #4261.

xbauch commented on regression/snapshot-harness/pointer_to_struct_01/test.desc:
```diff
@@ -5,8 +5,8 @@ st,p,q --harness-type initialise-with-memory-snapshot --initial-location main:4
-\[main.assertion.3\] line [0-9]+ assertion p != q: SUCCESS
-\[main.assertion.4\] line [0-9]+ assertion p == q: SUCCESS
+\[main.assertion.3\] line [0-9]+ assertion p \!= q: SUCCESS
+\[main.assertion.4\] line [0-9]+ assertion p == q-\>next: SUCCESS
```
This is unrelated but now all tests are green.

xbauch commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -191,6 +193,43 @@ exprt symbol_analyzert::get_char_pointer_value(
+  for(size_t i = 0; i < struct_type.components().size(); ++i)
+  {
+    const struct_typet::componentt &component = struct_type.components()[i];
+    const auto &bv_type = to_bitvector_type(component.type());
```
Obviously won't work if there is a non-bitvector member. But at least it will fail gracefully.

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.h:
```diff
@@ -145,6 +145,11 @@ class symbol_analyzert
     const exprt &zero_expr,
     const source_locationt &location);
 
+  exprt get_pointer_to_member_value(
```
Needs documentation, especially w.r.t. to the struct+offset thing

xbauch replied:
Done.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -0,0 +1,456 @@
+  }
+}
+
+exprt symbol_analyzert::get_pointer_to_member_value(
```
Could you actually make use of `get_subexpression_at_offset` to simplify this code?

xbauch replied:
Thanks, this is actually useful for unions as well.

martin-cs commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -200,6 +200,23 @@ void memory_snapshot_harness_generatort::add_init_section(
+{
+  symbolt &tmp_symbol = get_fresh_aux_symbol(
+    snapshot_symbol.type,
+    "",
```
???

martin-cs reviewed (commented):
This looks OK for now but can we get #4261 merged first?

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 2369a9b72125d788af181cb4ca36353953b1c453).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/109067689
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 24587e63e2b475b80595ef54763e59b330068985).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/109068469
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -167,6 +170,89 @@ exprt gdb_value_extractort::get_char_pointer_value(
+    const auto &value_symbol_expr = value_pair.second;
+    if(
+      struct_name ==
+      id2string(to_symbol_expr(value_symbol_expr).get_identifier()))
```
I don't think you need `id2string` here, but I could be wrong.

xbauch replied:
Had to swap the operands, but yes. Done.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -167,6 +170,89 @@ exprt gdb_value_extractort::get_char_pointer_value(
+  {
+    const typet target_type = expr.type().subtype();
+
+    symbol_exprt dummy(expr.type());
```
Missing name (this constructor is deprecated)

xbauch replied:
Added.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -167,6 +170,89 @@ exprt gdb_value_extractort::get_char_pointer_value(
+
+    values[memory_location] = new_symbol;
+
+    const auto &struct_symbol = values.find(memory_location);
```
Please use `insert` or `emplace` in the preceding instruction and then pick up the iterator being returned in order to avoid the repeated log-time lookup.

xbauch replied:
Done.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -167,6 +170,89 @@ exprt gdb_value_extractort::get_char_pointer_value(
+  if(maybe_member_expr.has_value())
+    return *maybe_member_expr;
+
+  UNREACHABLE;
```
Maybe instead use `CHECK_RETURN` above?

xbauch replied:
Done.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -212,6 +298,24 @@ exprt gdb_value_extractort::get_non_char_pointer_value(
+bool gdb_value_extractort::points_to_member(
+  const pointer_valuet &pointer_value) const
+{
+  if(pointer_value.pointee.find("+") != std::string::npos)
```
I'm not sure what the exact performance difference is, but intuitively I'd use `find('+')` instead of the string version to avoid any need for computing string lengths or the likes. Worth reviewing the entire PR for, this does occur multiple times.

xbauch replied:
Did you mean `std::find` or something internal? I use it twice so I added a wrapper method into `pointer_valuet` with a nice name.

tautschnig replied:
Right here I was thinking about `string::find`, but I could be wrong about the other cases. With `string::find` I'd just (when possible) actually use single characters rather than one-character strings.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -212,6 +298,24 @@ exprt gdb_value_extractort::get_non_char_pointer_value(
+  const symbolt *pointee_symbol = symbol_table.lookup(pointer_value.pointee);
+  if(pointee_symbol == nullptr)
+    return false;
+  const auto pointee_type = pointee_symbol->type;
```
`const auto &`

xbauch replied:
Done.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -212,6 +298,24 @@ exprt gdb_value_extractort::get_non_char_pointer_value(
+    pointee_type.id() == ID_array || pointee_type.id() == ID_struct ||
+    pointee_type.id() == ID_union)
+    return true;
+  return false;
```
I'd just use `return pointee_type.id() == ID_struct_tag || ...;`

xbauch replied:
Done.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -466,6 +470,39 @@ exprt gdb_value_extractort::get_struct_value(
+  const union_tag_typet &union_tag_type = to_union_tag_type(expr.type());
+  const union_typet union_type = ns.follow_tag(union_tag_type);
+
+  for(size_t i = 0; i < new_expr.operands().size(); ++i)
```
A union expression/union zero-initializer must always have exactly one operand.

xbauch replied:
Ok, I check that there is one and skip the loop.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -294,7 +294,9 @@ exprt gdb_value_extractort::get_non_char_pointer_value(
   {
-    return it->second;
+    const symbol_exprt typed_symbol_value = symbol_exprt{
+      to_symbol_expr(it->second).get_identifier(), expr.type().subtype()};
```
Possibly related comment on another commit or other PR (sorry, been reviewing too many PRs in too little time :-) ): can't we just store the symbol with the proper type so that `it->second` is fully typed?

xbauch replied:
Well I guess that part of the problem is that we initialise the `values` map before we know the types of the symbol expressions we store there. I only introduced this change because these initial symbol expressions do not have to have the right type. I guess we could get the right type during initialisation from symbol table but that change should probably take place in #4261.

tautschnig replied:
Ok, not a big deal, was just wondering.

petr-bauch committed:
Add building fresh symbols for snapshot temporaries

This is needed for example when referring to malloc(ed) variables. These are
build as pointing to `tmp` variables.

petr-bauch committed:
Add helper for querying offset

to pointer_valuet.

petr-bauch committed:
Make pointers-to-member work

Change the memory-analyze to produce correct member expressions. Iterate over
the offset from gdb, e.g. `st+4` means jump to the second member component
inside `st`.

petr-bauch committed:
Implement union values

Reuse get_struct_member for unions (copy-paste).

petr-bauch committed:
Add and check type of pointees in memory analyzer

GDB does not know (the CBMC) types but the resulting symbol table requires them.
We only do this for non-nil ireps (nil ireps can result from getting pointer
values for dynamically allocated memory -- will be addressed in later PRs).
Also prevents catch-by-value error when compiled with gcc-8.

petr-bauch committed:
Change the expected output of a regression test

but I think is a bit more meaningful now. We do not actually see the value of
`p->next` but at least it is consistent (before `p->next` was different from
`tmp.next`).

petr-bauch committed:
Add regression tests for snapshot-harness

These call memory-analyzer to produce the JSON snapshots and then build the
harness similarly to how goto-harness tests work. We currently tests structs,
unions, arrays and pointers to all of those.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: a68b10920d39970285e6f7d408775466030ec1bb).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/112107536
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e961c89b9f14fbc0acd1c61948615031c2017e87).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/112111854
